### PR TITLE
Fix stack overflow when attaching elements in maps

### PR DESCRIPTION
**Pull request by tederis**
When storing elements in maps sometimes we are attaching elements. In that case cross referencing may lead to 
an endless cycle and finally to a stack overflow error. The PR is intended to fix it.

The resource for crashing your own server:
[test.zip](https://github.com/multitheftauto/mtasa-blue/files/5212644/test.zip)

### CODE REVIEW

File: Client/mods/deathmatch/logic/CClientEntity.cpp [hunk 1]
```diff
@@ -1070,6 +1070,29 @@ bool CClientEntity::IsEntityAttached(CClientEntity* pEntity)
...
+
+        for (CClientEntity* pCurrent = this; pCurrent; pCurrent = pCurrent->GetAttachedTo())
+        {
+            if (pCurrent == pEntity)
+                return true;
+
+            if (MapContains(history, pCurrent))
+                break; // This should not be possible, but you never know
+
+            MapInsert(history, pCurrent);
```
Comment (Pirulax) [inline Client/mods/deathmatch/logic/CClientEntity.cpp]: Why not just use the return value of set::insert? It returns false if theres already a matching element in the set. Much faster than checking, then inserting.
Like this: 
```cpp
if (!std::get<bool>(history.insert(pCurrent)); 
    break; // This should not be possible, but you never know
```

File: Server/mods/deathmatch/logic/CElement.cpp [hunk 2]
```diff
@@ -1126,6 +1126,29 @@ bool CElement::IsElementAttached(CElement* pElement)
...
+
+        for (CElement* pCurrent = this; pCurrent; pCurrent = pCurrent->GetAttachedToElement())
+        {
+            if (pCurrent == pElement)
+                return true;
+
+            if (MapContains(history, pCurrent))
+                break; // This should not be possible, but you never know
+
+            MapInsert(history, pCurrent);
```
Comment (Pirulax) [inline Server/mods/deathmatch/logic/CElement.cpp]: Same suggestion here as before.

**Review by Pirulax — CHANGES_REQUESTED** [review]

File: Client/mods/deathmatch/logic/CClientEntity.cpp (hunk 1 above)
Comment (tederis) [reply Client/mods/deathmatch/logic/CClientEntity.cpp]: It is just a copy-paste from the old code

Comment (Pirulax) [reply Client/mods/deathmatch/logic/CClientEntity.cpp]: Yeah, I though, but its probably like a decade old, now there's a better way to do it :D

**Review by patrikjuvonen — APPROVED** [review]
Thank you for the pull request. Seems good to me.

**Review by ghost — APPROVED** [review]

**Comment by qaisjp** [comment]
My instinct is that we should (also|only) deal with this in the map parsing/loading process.

(I have only skimmed the code, so I don't know if this change affects performance, and therefore I don't know if the problem should ONLY be dealt with when loading maps.)

**Comment by tederis** [comment]
> My instinct is that we should (also|only) deal with this in the map parsing/loading process.
> 
> (I have only skimmed the code, so I don't know if this change affects performance, and therefore I don't know if the problem should ONLY be dealt with when loading maps.)

But it's already working that way. The processes are occuring in function LinkupElements() which in turn invoked inside a map loading process.

**Comment by qaisjp** [comment]
Great!

**Review by StrixG — APPROVED** [review]

**Pull request merged**